gossip/src/duplicate_shred.rs

```diff
@@ -84,35 +85,71 @@ pub enum Error {
     TryFromIntError(#[from] TryFromIntError),
     #[error("unknown slot leader: {0}")]
     UnknownSlotLeader(Slot),
+    #[error("invalid last index conflict")]
+    InvalidLastIndexConflict,
+    #[error("invalid erasure meta conflict")]
+    InvalidErasureMetaConflict,
 }
 
-// Asserts that the two shreds can indicate duplicate proof for
-// the same triplet of (slot, shred-index, and shred-type_), and
-// that they have valid signatures from the slot leader.
+/// Check that `shred1` and `shred2` indicate a valid duplicate proof
+///     - Must be for the same slot
+///     - Must have the same `shred_type`
+///     - Must both sigverify for the correct leader
+///     - If `shred1` and `shred2` share the same index they must be not equal
+///     - If `shred1` and `shred2` do not share the same index and are data shreds
+///       verify that they indicate an index conflict. One of them must be the
+///       LAST_SHRED_IN_SLOT, however the other shred must have a higher index.
+///     - If `shred1` and `shred2` do not share the same index and are coding shreds
+///       verify that they have conflicting erasure metas
 fn check_shreds<F>(leader_schedule: Option<F>, shred1: &Shred, shred2: &Shred) -> Result<(), Error>
 where
     F: FnOnce(Slot) -> Option<Pubkey>,
 {
     if shred1.slot() != shred2.slot() {
-        Err(Error::SlotMismatch)
-    } else if shred1.index() != shred2.index() {
-        // TODO: Should also allow two coding shreds with different indices but
-        // same fec-set-index and mismatching erasure-config.
-        Err(Error::ShredIndexMismatch)
-    } else if shred1.shred_type() != shred2.shred_type() {
-        Err(Error::ShredTypeMismatch)
-    } else if shred1.payload() == shred2.payload() {
-        Err(Error::InvalidDuplicateShreds)
-    } else {
-        if let Some(leader_schedule) = leader_schedule {
-            let slot_leader =
-                leader_schedule(shred1.slot()).ok_or(Error::UnknownSlotLeader(shred1.slot()))?;
-            if !shred1.verify(&slot_leader) || !shred2.verify(&slot_leader) {
-                return Err(Error::InvalidSignature);
-            }
+        return Err(Error::SlotMismatch);
+    }
+
+    if shred1.shred_type() != shred2.shred_type() {
+        return Err(Error::ShredTypeMismatch);
+    }
+
+    if shred1.index() == shred2.index() && shred1.payload() == shred2.payload() {
+        return Err(Error::InvalidDuplicateShreds);
+    }
+
+    if shred1.index() != shred2.index() && shred1.shred_type() == ShredType::Data {
+        match (
+            shred1.index(),
+            shred1.last_in_slot(),
+            shred2.index(),
+            shred2.last_in_slot(),
+        ) {
+            (ix1, true, ix2, false) if ix1 > ix2 => return Err(Error::InvalidLastIndexConflict),
+            (ix1, false, ix2, true) if ix1 < ix2 => return Err(Error::InvalidLastIndexConflict),
```

Comment:
What if both are `last_in_slot` but `ix1 != ix2`?

Reply:
It will get caught by the `_ => ()` case as this is a valid last index conflict proof.